### ISSUE TITLE
Should eslint throw error or auto fix missing necessary semi on fixing?

### ISSUE THREAD

**Issue by JounQin** [eslint/eslint#9354]
**Tell us about your environment**

* **ESLint Version:*4.7.2*
* **Node Version:*8.5.0*
* **npm Version:*5.4.2*

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```json
{
  "root": true,
  "env": {
    "browser": true,
    "node": true
  },
  "parserOptions": {
    "ecmaVersion": 2017
  },
  "extends": "eslint:recommended",
  "rules": {
    "no-console": 0
  }
}
```

</details>

**What did you do? Please include the actual source code causing the issue.**

```js
const a = 1(() => {
  console.log(a)
})()
```

**What did you expect to happen?**
add semi after `1`

**What actually happened? Please include the actual, raw output from ESLint.**
There is no semi after `1`.

**Comment by JounQin** [comment]
@platinumazure 
``` json
const a = 1(() => {
  console.log(a)
})()
```
this code passes through eslint, does that also work as intended?

**Comment by platinumazure** [comment]
Hi @JounQin, thanks for the issue.

Automatic semicolon insertion occurs with *most* line breaks, but not all. Basically, ASI does not occur if it is possible to join the lines and get valid syntax. In this case, yes, it is syntactically valid to call a function on a numeric literal.

I would highly recommend enabling the [`no-unexpected-multiline`](https://eslint.org/docs/rules/no-unexpected-multiline) rule. That rule is designed to catch newlines between tokens that look like the user might be assuming ASI is happening, but in reality it isn't. If you enable the rule in your original example, it should flag the issue.

**Comment by JounQin** [comment]
I use eslint with prettier, and I've enabled your recommended rule(eslint:recommended), so after I run `eslint --fix` with eslint, it will be rewrited as my above code, so I'm wondering would eslint be able to catch this problem.

**Comment by platinumazure** [comment]
@JounQin According to our demo, [`no-unexpected-multiline` should flag this case](https://eslint.org/demo/#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXVuZXhwZWN0ZWQtbXVsdGlsaW5lOiBlcnJvciAqL1xyXG5jb25zdCBhID0gMVxyXG4oKCkgPT4ge1xyXG4gIGNvbnNvbGUubG9nKGEpXHJcbn0pKCkiLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjYsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnt9LCJlbnYiOnt9fX0=). That rule does not autofix to remove line breaks (in ESLint).

I think this is an issue with prettier. Please file an issue there.

**Comment by JounQin** [comment]
@platinumazure you mean
``` js
const a = 1(() => {
  console.log(a)
})()
```
passes through eslint is also expected?

**Comment by platinumazure** [comment]
@JounQin Sure, how are we to know that the user doesn't actually want to do a function call on a numeric literal there?

`no-unexpected-multiline` specifically flags the newline case:

```js
const a = 1
(() => {
  console.log(a)
})()
```

ESLint does not transform the above example. Prettier is doing that. Prettier needs to take `no-unexpected-multiline` reports into account and not transform the line.

**Comment by platinumazure** [comment]
I'm going to go ahead and close this issue since Prettier is the one that did the transform, and since we recommended `no-unexpected-multiline` (covered under eslint:recommended) for catching the problem in the future.

Please reply here or open a new issue if you have further questions. Thanks!
